core/src/sumeragi/mod.rs

```diff
@@ -188,42 +178,60 @@ impl Sumeragi {
         block_hash: HashOf<VersionedCommittedBlock>,
     ) -> Vec<VersionedCommittedBlock> {
         self.internal
-            .wsv
+            .wsv_for_public_use
             .lock()
             .unwrap()
             .blocks_after_hash(block_hash)
     }
 
+    pub fn blocks_from_height(&self, block_height: usize) -> Vec<VersionedCommittedBlock> {
+        self.internal
+            .wsv_for_public_use
+            .lock()
+            .unwrap()
+            .blocks_from_height(block_height)
+    }
+
     pub fn get_random_peer_for_block_sync(&self) -> Option<Peer> {
-        use rand::{prelude::SliceRandom, SeedableRng};
+        use rand::{prelude::SliceRandom, RngCore, SeedableRng};
 
         let rng = &mut rand::rngs::StdRng::from_entropy();
-        self.internal
-            .wsv
+        let peers = self
+            .internal
+            .current_online_peers
             .lock()
-            .unwrap()
-            .peers()
-            .choose(rng)
-            .cloned()
+            .expect("lock on online peers for get random peer")
+            .iter()
+            .map(|peer| Peer::new((*peer).clone()))
+            .collect::<Vec<Peer>>();
+        if peers.is_empty() {
+            return None;
+        } else {
+            let mut sorted_peers = peers;
+            sorted_peers.sort();
+            Some(sorted_peers[rng.next_u32() as usize % sorted_peers.len()].clone())
+        }
     }
 
-    pub fn wsv_clone(&self) -> WorldStateView {
-        self.internal.wsv.lock().unwrap().clone()
+    pub fn wsv_mutex_access(&self) -> std::sync::MutexGuard<WorldStateView> {
+        self.internal.wsv_for_public_use.lock().unwrap()
     }
 
     pub fn initialize_and_start_thread(
         sumeragi: Arc<Self>,
         latest_block_hash: HashOf<VersionedCommittedBlock>,
         latest_block_height: u64,
     ) {
-        std::thread::spawn(move || {
-            fault::run_sumeragi_main_loop(
-                &sumeragi.internal,
-                latest_block_hash,
-                latest_block_height,
-            );
-            info!("Sumeragi Thread has Shutdown");
-        });
+        std::thread::Builder::new()
+            .name("sumeragi thread".to_string())
+            .spawn(move || {
+                fault::run_sumeragi_main_loop(
+                    &sumeragi.internal,
+                    latest_block_hash,
+                    latest_block_height,
+                );
+                info!("Sumeragi Thread has Shutdown");
```

Comment:
```suggestion
                info!("Sumeragi Thread Shutdown");
```
```suggestion
                info!("Sumeragi Thread has shut down");
```

Shutdown is a noun, if you want a verb, you should use "shut down"

Reply:
thank you